### ISSUE TITLE
[Feature Request] Add API to disable noqa support

### ISSUE THREAD

**Issue by plinss** [PyCQA/pydocstyle#484]
flake8 has an option to disable `# noqa` comments. When running pydocstyle as a flake8 plugin (via flake8-docstrings) and using flake8's `--disable-noqa` option, `# noqa` comments still block all violation reports from pydocstyle because pydocstyle is processing those comments itself and deciding to not send the reports at all.

It would be great if pydocstyle supported a feature equivalent to flake8's `--disable-noqa` and once that's in place, flake8-docstrings activated that feature so that flake8 sees all reports regardless of `# noqa` comments and could make its own decision to report them or not.

(I understand that would be a separate issue on flake8-docstrings once support is added to pydocstyle, I just mention it to ensure that the feature can be activated directly in `ConventionChecker.check_source()` as well as via traditional means).

**Comment by sigmavirus24** [comment]
Trying to hack this into flake8-docstrings will go poorly. As a Flake8 and Flake8-docstrings maintainer I'd rather start with something in the API only that allows us to gather all results because it will be very confusing going forward why we're overriding `--disable-noqa` for pydocstyle and that's not the good kind of maintenance burden.

**Comment by plinss** [comment]
I never proposed to hack something into flake8-docstrings to change pydocstyle's behavior, that *would* be painful to maintain, and it would be trying to solve the problem at the wrong layer. I did file this issue on pydocstyle, not flake8-docstrings.

I'm asking for feature parity between pydocstyle and flake8, they both parse and respect `# noqa` comments but only flake8 has the option to disable them (if pydocstyle has that ability I'm not finding it, happy to be proven wrong).

pydocstyle not having an equivalent to `--disable-noqa` (via any mechanism) is preventing flake8's `--disable-noqa` feature from working with pydocstyle violations. Getting that to work is my ultimate goal here, but pydocstyle getting a `--disable-noqa` option (or equivalent) for stand-alone use as well seemed reasonable to me.

If you'd rather expose this functionality in pydocstyle via an API only, or prefer some different way of getting violations from pydocstyle to flake8 that bypasses its `# noqa` handling that'd by fine by me.

**Comment by plinss** [comment]
To me it looks like adding a boolean argument (with default value matching current behavior, of course) to `ConventionChecker.check_source` and using that to optionally disable the `error.code not in definition.skipped_error_codes` test on line 124/125 would be the minimally invasive fix. 

e.g. 
```
--- src/pydocstyle/checker.py
+++ src/pydocstyle/checker.py
@@ -104,7 +104,7 @@ class ConventionChecker:
         ".+"            # Followed by 1 or more characters - which is the docstring for the parameter
     )
 
-    def check_source(self, source, filename, ignore_decorators=None):
+    def check_source(self, source, filename, ignore_decorators=None, all_errors=False):
         module = parse(StringIO(source), filename)
         for definition in module:
             for this_check in self.checks:
@@ -121,8 +121,8 @@ class ConventionChecker:
                         error = None
                     errors = error if hasattr(error, '__iter__') else [error]
                     for error in errors:
-                        if error is not None and error.code not in \
-                                definition.skipped_error_codes:
+                        if error is not None and (all_errors or error.code not in \
+                                definition.skipped_error_codes):
                             partition = this_check.__doc__.partition('.\n')
                             message, _, explanation = partition
                             error.set_context(explanation=explanation,
```
(deliberately didn't reformat to keep the diff small)

flake8-docstyle could then simply add `, all_errors=True` to it's call to get all errors and let flake8's `# noqa` handling take over from there.

Instead of a bool it could also be a list of codes to always return if you want more flexibility.

**Comment by sigmavirus24** [comment]
> flake8-docstyle could then simply add `, all_errors=True` to it's call to get all errors and let flake8's `# noqa` handling take over from there.

This is why I preferred this direction. flake8-docstrings really wants to give all of the errors to Flake8 and let that make decisions.

I think it's an anti-pattern for every tool to implement the same options as Flake8 and proliferate that maintenance across a bunch of already stretched thin maintainers. That's why I reacted to your solution-y feature request of adding more maintenance and support burden to this project

**Comment by plinss** [comment]
Apologies, I misstated my original ask. My primary concern was simply to enable this functionality for flake8-docstrings, which is why I explicitly wrote:

> I just mention it to ensure that the feature can be activated directly in ConventionChecker.check_source()

It just seemed to me that asking for API surface without exposing that to users was a bit odd, so I phrased it in the terms of a user-facing feature. Especially considering that apparently `# noqa` support was added to pydocstyle for feature parity with flake8, and it's not quite equivalent.

I'll modify the title of this issue to reflect just the API changes, and submit a separate issue to add the option to expose the API to users. Feel free to close that one as a "won't fix", but I think it's worth having the paper trail of the discussion in case it comes up again or others want to chime in.

In general, I agree that having `# noqa` processing in plugins to flake8 is an anti-pattern (and one I've seen in a few). My original thought was to request that it simply be dropped from pydocstyle, but I accept that pydocstyle is a stand-alone tool and people use it without flake8. So either internal `# noqa` processing or requiring some kind of wrapper to post-process the errors would be desired. This approach seems to serve both use cases.

While writing the tests, I noticed that pydocstyle also doesn't respect the flake8 `# noqa` feature of ignoring errors via prefix, e.g. `# noqa: D4`. I'll file a separate issue for that, again feel free to close as "won't fix".
